cura/Machines/QualityManager.py

```diff
@@ -0,0 +1,474 @@
+                    # into the current variant node.
+                    variant_node.addQualityMetadata(quality_type, metadata)
+                else:
+                    # If only variant_name and material are both specified, go one level deeper: create a material node
+                    # under the current variant node, and then add the quality/quality_changes metadata into the
+                    # material node.
+                    if root_material_id not in variant_node.children_map:
+                        variant_node.children_map[root_material_id] = QualityNode()
+                    material_node = variant_node.children_map[root_material_id]
+
+                    material_node.addQualityMetadata(quality_type, metadata)
+
+            else:
+                # If variant_name is not specified, check if material is specified.
+                if root_material_id is not None:
+                    if root_material_id not in machine_node.children_map:
+                        machine_node.children_map[root_material_id] = QualityNode()
+                    material_node = machine_node.children_map[root_material_id]
+
+                    material_node.addQualityMetadata(quality_type, metadata)
+
+        # Initialize the lookup tree for quality_changes profiles with following structure:
+        # <machine> -> <quality_type> -> <name>
+        quality_changes_metadata_list = self._container_registry.findContainersMetadata(type = "quality_changes")
+        for metadata in quality_changes_metadata_list:
+            if metadata["id"] == "empty_quality_changes":
+                continue
+
+            machine_definition_id = metadata["definition"]
+            quality_type = metadata["quality_type"]
+
+            if machine_definition_id not in self._machine_quality_type_to_quality_changes_dict:
+                self._machine_quality_type_to_quality_changes_dict[machine_definition_id] = QualityNode()
+            machine_node = self._machine_quality_type_to_quality_changes_dict[machine_definition_id]
+            machine_node.addQualityChangesMetadata(quality_type, metadata)
+
+        Logger.log("d", "Lookup tables updated.")
+        self.qualitiesUpdated.emit()
+
+    def _updateMaps(self):
+        self.initialize()
+
+    def _onContainerMetadataChanged(self, container):
+        self._onContainerChanged(container)
+
+    def _onContainerChanged(self, container):
+        container_type = container.getMetaDataEntry("type")
+        if container_type not in ("quality", "quality_changes"):
+            return
+
+        # update the cache table
+        self._update_timer.start()
+
+    # Updates the given quality groups' availabilities according to which extruders are being used/ enabled.
+    def _updateQualityGroupsAvailability(self, machine: "GlobalStack", quality_group_list):
+        used_extruders = set()
+        # TODO: This will change after the Machine refactoring
```

Comment:
I will change those marks. Those are more like remainders instead of TODOs. Since we have split the refactoring into multiple parts, I added those as remainders.

Reply:
But the machine refactoring is done now, right? It's even been merged into master.